readthedocs/v3/serializers.py

```diff
@@ -0,0 +1,444 @@
+            if code == programming_language:
+                return name
+        return 'Unknown'
+
+
+class ProjectURLsSerializer(serializers.Serializer):
+    documentation = serializers.CharField(source='get_docs_url')
+    project = serializers.SerializerMethodField()
+
+    def get_project(self, obj):
+        # Overridden only to return ``None`` when the description is ``''``
+        return obj.project_url or None
+
+
+class RepositorySerializer(serializers.Serializer):
+
+    url = serializers.CharField(source='repo')
+    type = serializers.CharField(source='repo_type')
+
+
+class ProjectLinksSerializer(BaseLinksSerializer):
+
+    _self = serializers.SerializerMethodField()
+
+    users = serializers.SerializerMethodField()
+    versions = serializers.SerializerMethodField()
+    builds = serializers.SerializerMethodField()
+    subprojects = serializers.SerializerMethodField()
+    superproject = serializers.SerializerMethodField()
+    translations = serializers.SerializerMethodField()
+
+    def get__self(self, obj):
+        path = reverse('projects-detail', kwargs={'project_slug': obj.slug})
+        return self._absolute_url(path)
+
+    def get_users(self, obj):
+        path = reverse(
+            'projects-users-list',
+            kwargs={
+                'parent_lookup_projects__slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+    def get_versions(self, obj):
+        path = reverse(
+            'projects-versions-list',
+            kwargs={
+                'parent_lookup_project__slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+    def get_builds(self, obj):
+        path = reverse(
+            'projects-builds-list',
+            kwargs={
+                'parent_lookup_project__slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+    def get_subprojects(self, obj):
+        path = reverse(
+            'projects-subprojects',
+            kwargs={
+                'project_slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+    def get_superproject(self, obj):
+        path = reverse(
+            'projects-superproject',
+            kwargs={
+                'project_slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+    def get_translations(self, obj):
+        path = reverse(
+            'projects-translations',
+            kwargs={
+                'project_slug': obj.slug,
+            },
+        )
+        return self._absolute_url(path)
+
+
+class ProjectSerializer(FlexFieldsModelSerializer):
+
+    language = LanguageSerializer()
+    programming_language = ProgrammingLanguageSerializer()
+    repository = RepositorySerializer(source='*')
+    privacy_level = PrivacyLevelSerializer(source='*')
+    urls = ProjectURLsSerializer(source='*')
+    subproject_of = serializers.SerializerMethodField()
```

Comment:
This and `translation_of` probably need to be prefetched in the queryset. I'm seeing a lot of duplicate queries to `projects_projectrelationship`.